File: ruby/triangle/triangle.rb
class Triangle

  def initialize(sides)
    @sides = sides
  end

  def triangle_inequality?
    sum = @sides.sum

    @sides.all? {|side| sum - side >= side }
  end

  def equilateral?
    @sides.uniq.length == 1 && @sides.none? { |side| side == 0 }
  end

  def two_sides_equal?
    @sides.uniq.length <= 2
  end

  def isosceles?
    triangle_inequality? && two_sides_equal?
  end

  def scalene?
    @sides.uniq.length == 3 && triangle_inequality?
  end

end

File: javascript/triangle/triangle.js
//
// This is only a SKELETON file for the 'Triangle' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export class Triangle {
  constructor(a, b, c) {
    this.sides = [a, b, c];
  };

  kind() {
    if (this.sides.some(side => side <= 0)) {
      throw 'sides must be greater than zero'
    }

    if (this.sides[0] + this.sides[1] < this.sides[2] ||
        this.sides[1] + this.sides[2] < this.sides[0] ||
        this.sides[2] + this.sides[0] < this.sides[1]
       ) {
      throw 'violates triangle inequality rule'
    }

    const side_count = (new Set(this.sides)).size;
    return side_count === 1 ? 'equilateral'
      : side_count === 2 ? 'isosceles'
      : 'scalene';
  }
}
